### ISSUE TITLE
After fixes were automatically applied the compiler reported errors within the file.

### ISSUE THREAD

**Issue by BenalexCheung** [rust-lang/rust#87786]
I tried this code:

```rust
use tokio::pin; // tokio = { version = "0.2", features = ["full"] }
use std::io;

async fn my_async_fn()  -> io::Result<()> {
    // async logic here
    Ok(())
}

#[tokio::main]
async fn main() {
    let future = my_async_fn();
    pin!(future);

    // (&mut future).await;
}
```

I expected to see this happen: *cargo fix*

Instead, this happened: 
```sh
warning: failed to automatically apply fixes suggested by rustc to crate `test_cargo_fix`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0425]: cannot find value `_future` in this scope
  --> src/main.rs:12:10
   |
12 |     pin!(_future);
   |          ^^^^^^^ help: a local variable with a similar name exists: `future`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0425`.
Original diagnostics will follow.

warning: unused variable: `future`
  --> src/main.rs:12:10
   |
12 |     pin!(future);
   |          ^^^^^^ help: if this is intentional, prefix it with an underscore: `_future`
   |
   = note: `#[warn(unused_variables)]` on by default
```

### Meta

`rustc --version --verbose`:
```
rustc 1.53.0 (53cb7b09b 2021-06-17)
binary: rustc
commit-hash: 53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b
commit-date: 2021-06-17
host: x86_64-unknown-linux-gnu
release: 1.53.0
LLVM version: 12.0.1
```

<details><summary>Backtrace</summary>
<p>

```sh
   Compiling test_cargo_fix v0.1.0 (/home/ben/project/test_cargo_fix)
warning: unused variable: `future`
  --> src/main.rs:12:10
   |
12 |     pin!(future);
   |          ^^^^^^ help: if this is intentional, prefix it with an underscore: `_future`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 1.23s
```

</p>
</details>

**Comment by BenalexCheung** [comment]
How about fixing lines 11 and 12 at the same time?

**Comment by SkiFire13** [comment]
MCVE:

```rust
macro_rules! foo {
    ($bind:ident) => {
        let $bind = &$bind;
    }
}

fn main() {
    let bar = 42;
    foo!(bar);
}
```

The problem seems to be that the compiler suggest changing an identifier based on whether a variable with that identifier is never used, ignoring however if that exact same identifier is used somewhere else (which can happen in a macro).

Possible solutions:
1. Suggest also changing all the other uses of that identifier. This has the downside that it will mark a variable that are actually used as non-used.
2. Don't suggest any change in case the same identifier is used somewhere else. (I prefer this one)

**Comment by BenalexCheung** [comment]
> MCVE:
> 
> ```rust
> macro_rules! foo {
>     ($bind:ident) => {
>         let $bind = &$bind;
>     }
> }
> 
> fn main() {
>     let bar = 42;
>     foo!(bar);
> }
> ```
> 
> The problem seems to be that the compiler suggest changing an identifier based on whether a variable with that identifier is never used, ignoring however if that exact same identifier is used somewhere else (which can happen in a macro).
> 
> Possible solutions:
> 
> 1. Suggest also changing all the other uses of that identifier. This has the downside that it will mark a variable that are actually used as non-used.
> 2. Don't suggest any change in case the same identifier is used somewhere else. (I prefer this one)

I Agree
